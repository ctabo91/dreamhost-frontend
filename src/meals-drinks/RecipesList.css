.filter-button {
    background-color: rgb(32, 26, 26);
    border: none;
    color: rgb(110, 66, 66);
}

.filter-button:hover {
    background-color: rgb(126, 37, 37);
    color: rgb(32, 26, 26);
}

.RecipesList-title {
    font-family: "Gruppo", sans-serif;
    font-weight: 400;
    font-style: normal;
    font-size: 50px;
}