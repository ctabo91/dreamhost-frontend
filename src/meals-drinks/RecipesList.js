import React, { useState, useEffect, useContext } from "react";
import { useParams, Link, useNavigate } from "react-router-dom";
import SearchForm from "../common/SearchForm";
import DreamHostApi from "../api/api";
import RecipeCard from "./RecipeCard";
import LoadingSpinner from "../common/LoadingSpinner";
import UserContext from "../auth/UserContext";
import "./RecipesList.css";

/** Shows a list of recipes, depending on the following factors:
 * - item === "meals" (list of all meals)
 * - item === "drinks" (list of all drinks)
 * - item === "meals" && category !== undefined (list of all meals in the specific category)
 * - item === "drinks" && category !== undefined (list of all drinks in the specific category)
 * 
 * The link to go to favorited recipes (either meals or drinks) is also shown here,
 * along with a link to the category list for either meals or drinks.
 */ 

function RecipesList() {
    const { currentUser, favMealIds, favDrinkIds } = useContext(UserContext);
    const { item, category } = useParams();
    const [recipes, setRecipes] = useState(null);
    const [favorites, setFavorites] = useState(false);
    const navigate = useNavigate();

    useEffect(function getRecipesOnMount() {
        category
                ? category === "personal"
                        ? searchPersonal(currentUser.username, item)
                        : search(item, {category})
                : search(item);
        
    }, [category, item, currentUser.username]);

    const searchPersonal = async (username, type) => {
        let personalRecipes = await DreamHostApi.getPersonalRecipes(username, type);
        setRecipes(personalRecipes);
    };

    const search = async (type, searchData) => {
        let recipes = await DreamHostApi.getRecipes(type, searchData);
        setRecipes(recipes);
    };

    const toggleFavorites = async () => {
        if (favorites && category) {
            navigate(`/recipes/${item}`);
        } else if (favorites && !category) {
            search(item);
        } else {
            await showFavRecipes();
        }

        setFavorites(!favorites);
    };

    const showFavRecipes = async () => {
        let favRecipes;
    
        if (item === "meals") {
            favRecipes = favMealIds ? await Promise.all([...favMealIds].map(async id => {
                return await DreamHostApi.getRecipe(id.toString(), item);
            })) : [];
        } else {
            favRecipes = favDrinkIds ? await Promise.all([...favDrinkIds].map(async id => {
                return await DreamHostApi.getRecipe(id.toString(), item);
            })) : [];
        }
    
        setRecipes(favRecipes);
    };
    
    const capitalizedItem = item[0].toUpperCase() + item.slice(1);

    if (!recipes) return <LoadingSpinner />;

    return (
        <div className="RecipesList col-md-8 offset-md-2 mt-5 pt-5">
            <h4 className="RecipesList-title display-4 text-center">
                {category && !favorites
                        ? `- ${category.toUpperCase()} RECIPES -`
                        : `- ${item.toUpperCase()} -`}
            </h4>
            {!category
                    ? <SearchForm searchFor={search} item={item} />
                    : null}
            
            {category !== "personal" 
                ? <div className="text-center">
                    <Link to={`/recipes/${item}/categories`}>
                        <button className="filter-button btn btn-sm btn-secondary mr-2">
                            Filter by Category
                        </button>
                    </Link>

                    <button className="filter-button btn btn-sm btn-secondary" onClick={toggleFavorites}>
                        {favorites ? `Back to All ${capitalizedItem}` : "Show Favorites"}
                    </button>
                  </div>
                : null}
            
            {recipes.length
                ? (
                    <div className="RecipesList-list mt-4">
                        {recipes.map(r => (
                            <RecipeCard 
                                key={r.id}
                                id={r.id}
                                name={r.name}
                                area={r.area}
                                type={r.type}
                                thumbnail={r.thumbnail}
                                ingredients={r.ingredients}
                                item={item}
                                access={category === "personal"
                                                ? "personal"
                                                : "global"}
                            />
                        ))}
                    </div>
                ) : (
                    <p className="lead">Sorry, no results were found!</p>
                )}
        </div>
    );
}

export default RecipesList;