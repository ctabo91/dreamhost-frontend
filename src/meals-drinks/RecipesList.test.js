import React from 'react';
import { render, screen, act, fireEvent } from '@testing-library/react';
import { MemoryRouter, Route, Routes } from 'react-router-dom';
import RecipesList from './RecipesList';
import UserContext from '../auth/UserContext';
import DreamHostApi from '../api/api';

jest.mock('../api/api');

const mockUserContext = {
  currentUser: {
    username: 'testUser',
  },
  favMealIds: [1],
  favDrinkIds: [1],
};

const mockMealRecipes = [
  { 
    id: 1,
    name: 'Recipe 1',
    area: 'Area 1',
    ingredients: ['Ingredient 1', 'Ingredient 2'],
    thumbnail: 'Thumbnail 1'
  }
];

const mockDrinkRecipes = [
  { 
    id: 1,
    name: 'Recipe 1',
    type: 'Type 1',
    ingredients: ['Ingredient 1', 'Ingredient 2'],
    thumbnail: 'Thumbnail 1'
  }
];

beforeEach(() => {
  jest.clearAllMocks();
});

describe('RecipesList Component', () => {
  it('matches snapshot', async () => {
    DreamHostApi.getRecipes.mockResolvedValueOnce(mockMealRecipes);

    const asFragment = await act(async () => {
      const { asFragment } = render(
        <UserContext.Provider value={mockUserContext}>
          <MemoryRouter initialEntries={['/recipes/meals']}>
            <Routes>
              <Route path="/recipes/:item" element={<RecipesList />} />
            </Routes>
          </MemoryRouter>
        </UserContext.Provider>
      );
      return asFragment;
    });

    expect(asFragment()).toMatchSnapshot();
  });

  it('renders loading spinner when recipes are null', async () => {
    await act(async () => {
      render(
        <UserContext.Provider value={mockUserContext}>
          <MemoryRouter initialEntries={['/recipes/meals']}>
            <Routes>
              <Route path="/recipes/:item" element={<RecipesList />} />
            </Routes>
          </MemoryRouter>
        </UserContext.Provider>
      );
    });

    expect(screen.getByText('Loading ...')).toBeInTheDocument();
  });

  it('renders meal recipes list when item is "meals"', async () => {
    DreamHostApi.getRecipes.mockResolvedValueOnce(mockMealRecipes);

    await act(async () => {
      render(
        <UserContext.Provider value={mockUserContext}>
          <MemoryRouter initialEntries={['/recipes/meals']}>
            <Routes>
              <Route path="/recipes/:item" element={<RecipesList />} />
            </Routes>
          </MemoryRouter>
        </UserContext.Provider>
      );
    });

    expect(screen.getByText('RECIPE 1')).toBeInTheDocument();
    expect(screen.getByText('AREA 1')).toBeInTheDocument();
  });

  it('renders drink recipes list when item is "drinks"', async () => {
    DreamHostApi.getRecipes.mockResolvedValueOnce(mockDrinkRecipes);

    await act(async () => {
      render(
        <UserContext.Provider value={mockUserContext}>
          <MemoryRouter initialEntries={['/recipes/drinks']}>
            <Routes>
              <Route path="/recipes/:item" element={<RecipesList />} />
            </Routes>
          </MemoryRouter>
        </UserContext.Provider>
      );
    });

    expect(screen.getByText('RECIPE 1')).toBeInTheDocument();
    expect(screen.getByText('TYPE 1')).toBeInTheDocument();
  });

  it('shows "Filter by Category" button', async () => {
    DreamHostApi.getRecipes.mockResolvedValueOnce(mockMealRecipes);

    await act(async () => {
      render(
        <UserContext.Provider value={mockUserContext}>
          <MemoryRouter initialEntries={['/recipes/meals']}>
            <Routes>
              <Route path="/recipes/:item" element={<RecipesList />} />
            </Routes>
          </MemoryRouter>
        </UserContext.Provider>
      );
    });

    expect(screen.getByRole('button', { name: /Filter by Category/i })).toBeInTheDocument();
  });

  it('toggles between show favorites and back to all recipes', async () => {
    DreamHostApi.getRecipes.mockResolvedValueOnce(mockMealRecipes);

    await act(async () => {
      render(
        <UserContext.Provider value={mockUserContext}>
          <MemoryRouter initialEntries={['/recipes/meals']}>
            <Routes>
              <Route path="/recipes/:item" element={<RecipesList />} />
            </Routes>
          </MemoryRouter>
        </UserContext.Provider>
      );
    });
    
    const showFavoriteButton = screen.getByRole('button', { class: 'filter-button-favorite', name: /Show Favorites/i });

    // Click the "Show Favorites" button
    DreamHostApi.getRecipe.mockResolvedValueOnce(mockMealRecipes[0]);
    await act(async () => {
      fireEvent.click(showFavoriteButton);
    });

    // Check that it shows the favorites
    expect(screen.getByText('RECIPE 1')).toBeInTheDocument();

    // Click the "Back to All Meals" button
    DreamHostApi.getRecipes.mockResolvedValueOnce(mockMealRecipes);
    await act(async () => {
      fireEvent.click(showFavoriteButton);
    });

    // Check that it shows all recipes
    expect(screen.getByText('RECIPE 1')).toBeInTheDocument();
  });
});
