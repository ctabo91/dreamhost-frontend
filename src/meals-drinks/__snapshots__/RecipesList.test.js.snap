// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RecipesList Component matches snapshot 1`] = `
<DocumentFragment>
  <div
    class="RecipesList col-md-8 offset-md-2 mt-5 pt-5"
  >
    <h4
      class="RecipesList-title display-4 text-center"
    >
      - MEALS -
    </h4>
    <div
      class="SearchForm mb-3"
    >
      <form>
        <div
          class="input-group"
        >
          <input
            class="form-control form-control-lg flex-grow-1"
            name="searchTerm"
            placeholder="Enter search term..."
            value=""
          />
          <button
            class="btn btn-lg"
            type="submit"
          >
            Submit
          </button>
        </div>
      </form>
    </div>
    <div
      class="text-center"
    >
      <a
        href="/recipes/meals/categories"
      >
        <button
          class="filter-button filter-button-category btn btn-sm btn-secondary mr-2"
        >
          Filter by Category
        </button>
      </a>
      <button
        class="filter-button filter-button-favorite btn btn-sm btn-secondary"
      >
        Show Favorites
      </button>
    </div>
    <div
      class="RecipesList-list mt-4"
    >
      <a
        class="RecipeCard card mb-3"
        href="/recipes/details/global/meals/1"
      >
        <div
          class="row no-gutters align-items-center"
        >
          <div
            class="col-md-4"
          >
            <img
              alt="Recipe 1"
              class="card-img img-fluid"
              src="Thumbnail 1"
            />
          </div>
          <div
            class="col-md-8"
          >
            <div
              class="card-body pt-0 pb-0"
            >
              <h3
                class="RecipeCard-title card-title text-center pb-0"
              >
                RECIPE 1
              </h3>
              <hr
                class="top-line mt-0"
              />
              <h4
                class="RecipeCard-subheading card-text text-center"
              >
                <i>
                  AREA 1
                </i>
              </h4>
              <hr
                class="mt-0"
              />
              <h5
                class="RecipeCard-ingredients text-center"
              >
                Ingredients
              </h5>
              <div
                class="RecipeCard-list text-center "
              >
                <div
                  class="RecipeCard-ingredient"
                >
                  Ingredient 1
                </div>
                <div
                  class="RecipeCard-ingredient"
                >
                  Ingredient 2
                </div>
              </div>
            </div>
          </div>
        </div>
      </a>
    </div>
  </div>
</DocumentFragment>
`;
